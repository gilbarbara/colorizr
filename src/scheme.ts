import { MESSAGES } from '~/modules/constants';
import { invariant } from '~/modules/invariant';
import { isHex, isNamedColor, isString } from '~/modules/validators';

import convert from '~/convert';
import extractColorParts from '~/extract-color-parts';
import rotate from '~/rotate';
import { ColorType } from '~/types';

export type Scheme =
  | 'analogous'
  | 'complementary'
  | 'rectangle'
  | 'split'
  | 'split-complementary'
  | 'square'
  | 'tetradic'
  | 'triadic';

export interface SchemeOptions {
  format?: ColorType;
  /**
   * The type of scheme to generate.
   * @default 'complementary'
   */
  type?: Scheme;
}

/**
 * Get the scheme for a color.
 */
export default function scheme(input: string, typeOrOptions?: Scheme | SchemeOptions): string[] {
  invariant(isString(input), MESSAGES.inputString);
  const { format, type = 'complementary' } = isString(typeOrOptions)
    ? { type: typeOrOptions }
    : (typeOrOptions ?? {});

  const output = isHex(input) || isNamedColor(input) ? 'hex' : extractColorParts(input).model;

  const colors: string[] = [];

  switch (type) {
    case 'analogous': {
      colors.push(rotate(input, -30), input, rotate(input, 30));
      break;
    }
    case 'complementary': {
      colors.push(input, rotate(input, 180));
      break;
    }

    case 'split':
    case 'split-complementary': {
      colors.push(input, rotate(input, 150), rotate(input, 210));
      break;
    }
    case 'triadic': {
      colors.push(input, rotate(input, 120), rotate(input, 240));
      break;
    }

    case 'tetradic':
    case 'rectangle': {
      colors.push(input, rotate(input, 60), rotate(input, 180), rotate(input, 240));
      break;
    }
    case 'square': {
      colors.push(input, rotate(input, 90), rotate(input, 180), rotate(input, 270));
      break;
    }
    default: {
      throw new TypeError('invalid type');
    }
  }

  return colors.map(color => convert(color, format ?? output));
}
