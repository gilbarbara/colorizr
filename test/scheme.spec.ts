import scheme, { SchemeOptions } from '~/scheme';

import { brightPink, green, orange, violet, yellow } from './__fixtures__';

describe('scheme', () => {
  it.each<[{ input: string; options?: SchemeOptions }]>([
    [{ input: brightPink.hex, options: undefined }],
    [{ input: brightPink.hex, options: { type: 'complementary' } }],
    [{ input: green.hslString, options: { type: 'analogous' } }],
    [{ input: green.hslString, options: { type: 'split' } }],
    [{ input: orange.oklabString, options: { type: 'split-complementary' } }],
    [{ input: orange.oklabString, options: { type: 'triadic' } }],
    [{ input: violet.oklchString, options: { type: 'tetradic' } }],
    [{ input: violet.oklchString, options: { type: 'rectangle' } }],
    [{ input: yellow.rgbString, options: { type: 'square' } }],
    [{ input: 'AliceBlue', options: undefined }],
  ])('should return properly with $input and $options', ({ input, options }) => {
    expect(scheme(input, options)).toMatchSnapshot();
  });

  it('should fail with invalid parameters', () => {
    // @ts-expect-error - invalid parameters
    expect(() => scheme('#f04', 'invalidType')).toThrow('invalid type');
  });
});
