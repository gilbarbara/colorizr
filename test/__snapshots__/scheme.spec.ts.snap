// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`scheme > should return properly with '#ff0044' and { type: 'complementary' } 1`] = `
[
  "#ff0044",
  "#00ffbb",
]
`;

exports[`scheme > should return properly with '#ff0044' and undefined 1`] = `
[
  "#ff0044",
  "#00ffbb",
]
`;

exports[`scheme > should return properly with 'AliceBlue' and undefined 1`] = `
[
  "#f0f8ff",
  "#fff7f0",
]
`;

exports[`scheme > should return properly with 'hsl(136 100% 50%)' and { type: 'analogous' } 1`] = `
[
  "hsl(106 100% 50%)",
  "hsl(136 100% 50%)",
  "hsl(166 100% 50%)",
]
`;

exports[`scheme > should return properly with 'hsl(136 100% 50%)' and { type: 'split' } 1`] = `
[
  "hsl(136 100% 50%)",
  "hsl(286 100% 50%)",
  "hsl(346 100% 50%)",
]
`;

exports[`scheme > should return properly with 'oklab(70.622% 0.1374 0.14283)' and { type: 'split-complementary' } 1`] = `
[
  "oklab(70.622% 0.1374 0.14283)",
  "oklab(89.962% -0.15715 -0.01587)",
  "oklab(45.838% -0.03153 -0.30726)",
]
`;

exports[`scheme > should return properly with 'oklab(70.622% 0.1374 0.14283)' and { type: 'triadic' } 1`] = `
[
  "oklab(70.622% 0.1374 0.14283)",
  "oklab(87.269% -0.21232 0.12962)",
  "oklab(51.156% 0.09061 -0.27875)",
]
`;

exports[`scheme > should return properly with 'oklch(47.642% 0.29956 274.93693)' and { type: 'rectangle' } 1`] = `
[
  "oklch(47.642% 0.29956 274.93693)",
  "oklch(66.672% 0.28363 344.9015)",
  "oklch(92.006% 0.23635 126.69487)",
  "oklch(86.876% 0.27606 144.65534)",
]
`;

exports[`scheme > should return properly with 'oklch(47.642% 0.29956 274.93693)' and { type: 'tetradic' } 1`] = `
[
  "oklch(47.642% 0.29956 274.93693)",
  "oklch(66.672% 0.28363 344.9015)",
  "oklch(92.006% 0.23635 126.69487)",
  "oklch(86.876% 0.27606 144.65534)",
]
`;

exports[`scheme > should return properly with 'rgb(255 230 109)' and { type: 'square' } 1`] = `
[
  "rgb(255 230 109)",
  "rgb(109 255 157)",
  "rgb(109 134 255)",
  "rgb(255 109 207)",
]
`;
